### PR TITLE
ctools2 - Create view and remove manual tables

### PR DESCRIPTION
**Pull request by ccjjmartin**
### Purpose:
- Create one view with two displays to replace the manual tables created
### Steps to update:
- Run update.php to revert your iNope Policy Review feature.
### Staff procedures:
- Login as a Staff employee
- Navigate to /node/*/review and verify that the views block is appearing
- Also verify that you can only see comments by that user
- Try to edit a comment and verify form functionality is still working
- View a basic page and verify that the view is not appearing
### Admin procedures:
- Logout and Login as an Administrator
- Navigate to /node/*/review and verify that the views block is appearing
- Navigate to /admin/inope/report and verify that the views block is appearing
- Try to edit a comment and verify form functionality is still working (create a new revision if needed)
- View a basic page and verify that the view is not appearing

### CODE REVIEW

**Comment by thebruce** [comment]
I got this error when I tried the steps above:

Fatal error: Access level to inope_policy_review_handler_note_status_boolean::option_definition() must be public (as in class views_handler_field_boolean) in /var/www/aquifer-inope/modules/custom/inope_policy_review/inope_policy_review_handler_note_status_boolean.inc on line 18

File: modules/custom/inope_policy_review/inope_policy_review_handler_note_status_boolean.inc [hunk 1]
```diff
@@ -10,18 +10,21 @@
...
-  function option_definition() {
+  /**
+   * Override the default boolen field handlers default options.
+   *
+   * We are changing the default option to approved-requested created in init.
+   *
+   * @return array $options
+   *   Sets the default options to be used in the init method.
+   */
+  protected function option_definition() {
```
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review_handler_note_status_boolean.inc]: So this is the line that is causing the issue -- why did you choose protected here?

It's important to understand how implementing classes work with this PHP OOP stuff. If you are overriding existing classes - changing the scope can cause problems.

File: modules/custom/inope_policy_review/inope_policy_review_handler_note_status_boolean.inc [hunk 2]
```diff
@@ -31,11 +34,23 @@ class inope_policy_review_handler_note_status_boolean extends views_handler_fiel
...
+   * We are adding the approved-requested option.
+   *
+   * @param array $view
+   *    Todo - passed by reference.
+   * @param array $options
+   *    Todo - passed by reference.
+   *
+   * @todo research the parameters passed and update documentation.
+   */
+  protected function init(&$view, &$options) {
```
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review_handler_note_status_boolean.inc]: Why did you move towards protected again?  The default (no scope indicated) in php is public, I believe - unless there is something that you are trying to do here.

**Comment by thebruce** [comment]
Functional checks out -- time for code review!

**Comment by thebruce** [comment]
I think we had verbally talked about the views.inc file going either into an includes folder or includes/views - I'd still like to see that happen.

**Comment by thebruce** [comment]
Also the handler needs to go into includes/views/handlers or something like that -- 

It's good to keep this stuff organized in the module, otherwise it can get messy.

Additionally at a glance, the folder structure that has a views folder will say something to a developer immediately rather than a wall of files in a module -- which may or may not carry the same impact.

File: modules/custom/inope_policy_review/inope_policy_review.module [hunk 3]
```diff
@@ -173,7 +165,7 @@ function inope_policy_review_page($node) {
     // Build a table of results from other users.
     $build['table'] = array(
       '#type' => 'markup',
-      '#markup' => inope_policy_review_table($node, $account),
+      '#markup' => views_embed_view('policies', 'block_1'),
```
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review.module]: Is the render array overkill with views_embed_view?  What does the output look like either way?  If you can get away with only views embed view and the output isn't disgusting, that'd be cleaner I think.

File: modules/custom/inope_policy_review/inope_policy_review.module [hunk 4]
```diff
@@ -566,199 +542,6 @@ function inope_policy_review_update_query(&$form_state) {
 }
 
 /**
- * Retrieve admin table from cache if possible.
```
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review.module:569]: One thing I would hoped you would have brought up in the client conversation yesterday was that caching was going to be lost -- or that it would change - depending on what your plans are for that.

Then again looking at this -- the caching with the sorts was something we talked about redoing in our last demo - so it's probably ok to say goodbye to this particular implementation.

File: modules/custom/inope_policy_review/inope_policy_review.views.inc [hunk 5]
```diff
@@ -62,12 +39,32 @@ function inope_policy_review_views_data() {
...
   );
 
-  // Policy Review NID
+  // Create the policy review id (primary key) views field.
+  $data['policy_review']['policy_review_id'] = array(
+    // The item it appears as on the UI.
+    'title' => t('Policy Review ID'),
+
+    // The help that appears on the UI.
+    'help' => t('The primary key.'),
```
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review.views.inc]: The primary key for...?  Might be useful - presumably there are other primary keys out there.  Just a thought.

**Comment by thebruce** [comment]
A couple comments about the view - since we only talked about it and I didn't get a chance to do a review of it:

You actually created the view as a Policy node base table which is weird and is why you need all the relationships. It does work -- so cool. I think my preference would have been to start it as a node base view and add in the relationship to the policy review via node and perhaps user if necessary -- that would allow you to use Node access, which could be useful.  

I appreciate how you used the new base table -- how would this look using a node base table? Any problems getting the same results?

Block Display:
In the contextual Filter in the block can you apply some validation to the nid?

The admin can no longer see all input on the node/*/review page.  Do we need another display to enable this? 

It is best practice to provide a good relevant machine name -- block_1 is not great.

You have all of the relationships required - do you need them to be required? How are those values being determined? (EDIT - given the base table you selected - yes.  However - with a node base table -- would you need all of them?)

File: modules/custom/inope_policy_review/inope_policy_review.module (hunk 3 above)
Comment (ccjjmartin) [inline modules/custom/inope_policy_review/inope_policy_review.module]: views_embed_view returns HTML to be displayed on the page the render array is used as we have multiple items to display (either the form+view or message+view).

Do you have another method in mind you would like me to try?

File: modules/custom/inope_policy_review/inope_policy_review.module (hunk 4 above)
Comment (ccjjmartin) [inline modules/custom/inope_policy_review/inope_policy_review.module:569]: Caching ran through my mind briefly but I knew views had a caching option that we could reimplement.

**Comment by ccjjmartin** [comment]
I used the policy review base table because the solution was not meant to implement hook_views_data_alter() so I didn't change the node table. Also, I could have created a join for my policy base table which would allow other tables access to my table.  In reality when I first ran through this I was looking to attach it to the node table but, I was experiencing problems getting the file to be read through hook_view_api() this way worked so I didn't mess with it.

Using the node base table would look different in regards to the relationships used when adding fields to the view.  I think it would look cleaner and have additional functionality if it was using the node base table.  However, my understanding is that the base table is the table that other tables are joined to so joining to the node table would increase our base query size as the node table includes other content types.  This would result in a slower running query for information we don't really need (other content types).

We could add validation to the contextual filter.  I am not sure what the benefit would be (at least as it is now).  We currently only display the view through code views_embed_view on the pages which already have built in validation from the previous exercises.  Once we switch to panels that might break access checks but until then the current validation is done through:

```
1) hook_access_callback verifies the content type (inope_policy)
2) hook_menu verifies it is a node so it must be numeric
3) hook_access_callback verifies it is only displayed to users with the 'view inope policy review' permission

Did I miss an access check we should be doing?
```

I created a new display for admin on nodes created and is now being displayed, you will need to get the new code, clear your caches, and revert the inope_policy feature.

All machine names have been updated.

With a node base table we would have different relationships offered by the node table (node_revisions, policy_review, user) which would still be required as we would want to hide items without a policy_review relationship.  We could probably get away without node_revisions and user based on the idea that all nodes should have node_revisions and users so our check is additional overhead on those two.

File: modules/custom/inope_policy_review/inope_policy_review.module (hunk 4 above)
Comment (thebruce) [inline modules/custom/inope_policy_review/inope_policy_review.module:569]: it comes up a little bit later in the exercises, so there will be a space to think about it then -- however, I think i want to illustrate that when you are proposing a different or new solution - you kind of do need to hit a point to point, if this were released as working functionality - which is the intent (with a few weeks or so going by with this deployed) and the system had any kind of load or volume  you can bet the fact that there is no caching would be noticed.

**Comment by thebruce** [comment]
In response to node table --

OK -- yeah, this makes sense.  I think a good addition would be to write out a $data['policy_review']['table']['join'].  The default only supports a single join - yet - this would provide additional functionality and access to this information from a node base table view -- which is probably the most common kind of view anyway.

I think you've made a good justification for keeping these reports in the inope policy review base table.

The validation I was talking about on the contextual argument -- was in the UI -- did you attempt to use nid validation on the view?  You are right about what's happening in the form but for the sake of future proofing the view and best practice its good to put something in there or get used to doing that -- of course sometimes it breaks stuff when you do -- ha, views :dancer: 

I'd like to see the above Join put into the views data definition for the inope table, once you get that working feel free to merge and move ahead.

Good justification above.

**Comment by thebruce** [comment]
Render array is fine - i was thinking of some solutions a little further down the line in the theme and render space, but they would not actually be improvements.
